### DISCUSSION TITLE
How does shard rebalancing work in a cluster according to Deployment Strategy in Kubernetes?

### DISCUSSION THREAD

**Discussion by civilizeddev** [Tochemey/goakt#190] [Q&A] [answered]
Hi, @Tochemey 

Today, I have some questions about how the cluster and shards behave when there is a change in the members that make up the cluster.

There are mainly two deployment strategies that can be used to update apps in a cluster:

- Rolling Update
- Blue/Green Update

In the following case, when the cluster detects changes in membership and rebalances shards,

At such moments, it feels like there might be sudden fluctuations. I'm also curious if this leads to a chain of actor termination and respawn among shards.

![](https://www.plantuml.com/plantuml/svg/ZPIzReCm54PtFyLLkhH38k6VfOfGdg5LvyZW2q4rs2BEW53VlQHIbMlJ0faSVKU6SmPErw3xS6ijOzVFflEwrotOfiCmU0Jlx73_yzVc-d1nZ5ci0WG7VLDV0fYcnp8qhcFiVj1TRH5sxw7N0UlX5TwSdIweuSCR7N07OyGzPoyxq5SmwAqRMkn2DX401D7Ux_KFD9tEtfcppk1ud0vvFYUBY2mMo38YomV9W3vfJouZIq6NCcDViHPFjVW6BVxx2l3KerbruvlT5aWJaOQyWcrQPDIOKsDEbf8kcaS1EHmEERa_3INIK0B6LNEn72hL5vlr5qYCI1JpKEHvUt8bMGnT2X7L5L4e56aecOQICAwQo-LGgRxShBz0LX5PoVz36Kboe2GBqgMKKLqPXQfa6agbeHICg-PgixdQRFuWgREYpeewqwLIazOaHh-BX-cN4ozELZvR1Ld5R9LabRDLiHDsvjRQRm00)

![](https://www.plantuml.com/plantuml/svg/ZPMnpjem58LtFyKLNThXfuhjgK8Hwj87g3eZ4rz3L2UnZ1aYr7TlyWzmR109OZ3wFjkwvqXcVqacfcldnULlsmSJJKU-xJcDWIaCVfn_l693JET12CykKHeejgTp8jj6Rb8xz4BC2zEVFDFcTuecyMdyGJ_zbR__Yimz_GdM9DxGBLE_LjysP2vaEVXXxBXFrMGGWM92-32VqhGwXC4UJ4MxtRIewsVpc9d7DsQJcSt358GtVS1XG8v8cahyoyUIvLXonLXozLXozLXlJ9kPzcxI5zzs_EnpvlEYxpBVbJx6B35cYJ4ZQIHkieWODp4I9ukRjjiQpi2xSiR8cfmvMHQkoiBLYiBL8vB8Th7rN5-iFjSN-x_hYys3kT3vmyJY51QdW30IfxBYL5wSmk9oniYQd3bL5gVBuZJT5hFIgxFIgxFIBxFIc9K6maYSdiUQllYERkTFZMyGO4a5L03Lyrujn9vxE_zn_0S0)

Is the recommended deployment strategy in this case likely to be Blue/Green Update?

Thank you.

**Comment by Tochemey** [comment, marked as answer]
@civilizeddev Thanks for your question. Let me explain the current design:

- The cluster implementation simply relies on a distributed sharded key/value store [engine](https://github.com/buraksezer/olric). It handles the rebalance and the sharding mechanism.

- A single instance of an actor is created when cluster mode is enabled. The actor identifier is propagated across the cluster.
- Each node in the cluster has a copy of the created actor identifier and can easily refers to the location of the actor. Now that I am explaining this I just discover an issue :). When the node where the given actor is created dies, then the other nodes in the cluster are aware of the event. However, at the moment, Go-Akt is not doing actor redeployment which is the issue we need to address. It is an issue because that actor cannot be found when its node dies.

I hope I clarify the cluster implementation a bit.

I have created an issue for that https://github.com/Tochemey/goakt/issues/192

**Comment by civilizeddev** [comment]
@Tochemey 

In https://github.com/Tochemey/goakt/issues/192#issuecomment-1868574816

Are you saying that you're considering abandoning the use of [Olric](https://github.com/buraksezer/olric) as a cluster engine and looking into [Raft](https://github.com/etcd-io/raft) as an alternative? I'm not really familiar with either. What are the shortcomings that necessitate a replacement? If you could explain it simply, I think I could learn.

**Comment by Tochemey** [comment]
@civilizeddev
The current cluster implementation does work as expected. One can access actors from any node as long as the actor is alive or the node hosting the actor is not dead(a.k.a left the cluster). Therefore to tackle actors' redeployment when the hosting node is dead, I am saying I may consider other technologies or protocols in case Olric does not meet the required needs. This is still WIP.
[Olric](https://github.com/buraksezer/olric) is a distributed cluster in-memory store. You can read more on their readme.
Regarding the raft protocol, I will suggest this lovely [Wikipedia page](https://en.wikipedia.org/wiki/Raft_(algorithm)). In a nutshell, it is a protocol that helps build distributed and clustered-based applications.
